include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -325,13 +308,14 @@ struct SCookTorrance
 
         dg1_query_type dq = ndf.template createDG1Query<Interaction, MicrofacetCache>(interaction, cache);
 
-        fresnel_type _f = impl::getOrientedFresnel<fresnel_type, IsBSDF>::__call(fresnel, interaction.getNdotV());
+        fresnel_type _f = getOrientedFresnel(fresnel, interaction.getNdotV());
         quant_query_type qq = impl::quant_query_helper<ndf_type, fresnel_type, IsBSDF>::template __call<Interaction, MicrofacetCache>(ndf, _f, interaction, cache);
         quant_type DG1 = ndf.template DG1<sample_type, Interaction>(dq, qq, _sample, interaction, isInfinity);
 
         NBL_IF_CONSTEXPR(IsBSDF)
         {
-            const scalar_type reflectance = _f(hlsl::abs(cache.getVdotH()))[0];
+            spectral_type prefixThroughputWeights = interaction.getPrefixThroughputWeights();
+            const scalar_type reflectance = hlsl::dot(impl::__implicit_promote<spectral_type, typename fresnel_type::vector_type>::__call(_f(hlsl::abs(cache.getVdotH()))), prefixThroughputWeights);
```

Comment:
ok you don't want to promote a 1D value to 3D to do a dot product which you know weights to 1.f

You want to do an
```cpp
fresnel_type::vector_type fresnelValue = _f(hlsl::abs(cache.getVdotH()))[0];
NBL_IF_CONSTEXPR(fresnel_type::monochrome)
{
   reflectance = fresnelValue;
}
else
{
   reflectance = hlsl::dot(interaction.getPrefixThroughputWeights(),fresnelValue);
}
```

Also something somewhere (the Config) needs to check that the Fresnel `operator()`'s return type is either monochrome or the same as Interaction's `spectral_type`

Finally you probably need this commonalized into some pseudo-private method for the BSDF cook torrance, since you need it for:
- generate to decide reflection vs refraction
- pdf to compute the pdf
- quotient_and_pdf to compute both